driver/src/main/java/org/neo4j/driver/internal/InternalSession.java

```diff
@@ -121,6 +121,27 @@ public TypeSystem typeSystem()
         return InternalTypeSystem.TYPE_SYSTEM;
     }
 
+    @Override
+    protected void finalize() throws Throwable
+    {
+        if( isOpen )
+        {
+            // Using stderr here to avoid calling into a logging system that may not exist when this finalizer runs.
+            System.err.println( "WARNING: Neo4j Session object leaked, please ensure that your application calls the `close` method on Sessions before disposing of the objects." );
+            if( currentTransaction != null )
+            {
+                // If there is an open transaction, rather than deal with the complexity of cleaning that up inside the GC finalizer,
+                // force the connction to be closed, rather than returned to the connection pool.
+                // This can be removed once we have a good way in Bolt to reset a session, then we won't need to close it like this.
+                connection.forceClose();
+            }
+            else
+            {
+                connection.close();
+            }
+        }
+    }
```

Comment:
Good call. Also - WDYT about using the logging framework here rather than stderr? I'm kind of scared of finalizers, I was worried it'd risk leading down a rabbit hole of broken object graphs to call "too far out". Do you think it'd be safe to keep a `final Log` field on the Session and use that to print the warning?